fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,336 @@
+use serde::{Deserialize, Serialize};
+use std::fs::{self, File};
+use std::io::{Read, Write};
+use std::time::SystemTime;
+use std::{
+    collections::BTreeMap,
+    path::{Path, PathBuf},
+};
+
+use crate::base_storage::{BaseStorage, SyncStatus};
+use crate::monoid::Monoid;
+// use crate::utils::read_version_2_fs;
+use data_error::{ArklibError, Result};
+
+/*
+Note on `FolderStorage` Versioning:
+
+`FolderStorage` is a basic key-value storage system that persists data to disk.
+
+In version 2, `FolderStorage` stored data in a plaintext format.
+Starting from version 3, data is stored in JSON format.
+
+For backward compatibility, we provide a helper function `read_version_2_fs` to read version 2 format.
+*/
+const STORAGE_VERSION: i32 = 3;
+
+/// Represents a file storage system that persists data to disk.
+pub struct FolderStorage<K, V>
+where
+    K: Ord,
+{
+    /// Label for logging
+    label: String,
+    /// Path to the underlying file where data is persisted
+    path: PathBuf,
+    /// Last modified time of internal mapping. This becomes equal to
+    /// `written_to_disk` only when data is written or read from disk.
+    // modified: SystemTime,
+    ram_timestamps: BTreeMap<K, SystemTime>,
```

Comment:
We have three btrees here mapping the same key to different values. Have you considered merging these into a single tree that stores a single entry that contains the value, the ram timestamp and the disk timestamp? Investigate if this is a better approach.

Reply:
I tried it, but it makes the implementation less readable.